crates/floresta-wire/src/p2p_wire/tests/sync_node.rs

```diff
@@ -0,0 +1,170 @@
+        let mempool = Arc::new(RwLock::new(Mempool::new()));
+        let chain = ChainState::new(chainstore, network, AssumeValidArg::Disabled);
+        let chain = Arc::new(chain);
+
+        // Adding 9 signet headers in the chain-state prior validation
+        let mut headers = get_test_headers();
+        headers.remove(0);
+        headers.truncate(9);
+        for header in headers {
+            chain.accept_header(header).unwrap();
+        }
+
+        let config = get_node_config(datadir, network, pow_fraud_proofs);
+
+        let mut node = UtreexoNode::<SyncNode, Arc<ChainState<KvChainStore>>>::new(
+            config,
+            chain.clone(),
+            mempool,
+            None,
+        );
+
+        for (i, peer) in peers.into_iter().enumerate() {
+            let (sender, receiver) = async_std::channel::bounded(10);
+            let peer = create_peer(
+                peer.0,
+                peer.1,
+                peer.2,
+                node.node_tx.clone(),
+                sender.clone(),
+                receiver,
+                i as u32,
+            );
+
+            let _peer = peer.clone();
+
+            node.peers.insert(i as u32, peer);
+        }
+        let mut node = ManuallyDrop::new(Box::new(node));
+
+        let kill_signal = Arc::new(RwLock::new(false));
+        // FIXME: This doesn't look very safe, but we need to coerce a &mut reference of the node
+        //        to live for the static lifetime, or it can't be spawn-ed by async-std::task
+        let _node: &'static mut UtreexoNode<SyncNode, Arc<ChainState<KvChainStore>>> =
+            unsafe { std::mem::transmute(&mut **node) };
+
+        future::timeout(Duration::from_secs(10), _node.run(kill_signal, |_| {}))
+            .await
+            .unwrap();
+
+        chain
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use std::collections::HashMap;
+
+    use floresta_chain::pruned_utreexo::BlockchainInterface;
+    use floresta_chain::UtreexoBlock;
+
+    use crate::p2p_wire::tests::sync_node::tests_utils::setup_node;
+    use crate::p2p_wire::tests::utils::get_essentials;
+
+    #[async_std::test]
+    async fn test_sync_valid_blocks() {
+        let (headers, blocks, _, _, _) = get_essentials();
+        let chain = setup_node(
+            vec![(Vec::new(), blocks.clone(), HashMap::new())],
+            false,
+            floresta_chain::Network::Signet,
+        )
+        .await;
+
+        assert_eq!(chain.get_validation_index().unwrap(), 9);
+        assert_eq!(chain.get_best_block().unwrap().1, headers[9].block_hash());
+        assert_eq!(chain.is_in_idb(), false);
+    }
+
+    #[async_std::test]
+    async fn test_sync_invalid_block() {
+        // 7th BLOCK IS SET AS INVALID. WHILE CONNECTING THE BLOCKS, 7th BLOCK WILL BE INVALIDATED.
+        // HENCE THE CHAIN WILL HAVE A HEIGHT OF 6.
+
+        // THIS SIMULATION WILL TEST:
+        // 1) SENDING BLOCK WITH INVALID PROOF: 7TH BLOCK IS SET AS INVALID.
```

Comment:
Is the proof invalid, or the actual block? If the proof is invalid, we don't invalidate a block because this is not PoW committed and could easily be forged by peers.

Reply:
My bad here, I tweaked with the transactions so currently it is giving `BadMerkelRoot`, in the `self.validate_block` function in `connect_block`. So its invalidating the block.

Should I let it be and change the comments accordingly or ... ??

Reply:
But regardless the `InvalidProof` message, comes from the `validate_block` fucntion in the `BlockchainInterface` trait here: 
```rust
fn validate_block(
        &self,
        block: &Block,
        proof: Proof,
        inputs: HashMap<OutPoint, TxOut>,
        del_hashes: Vec<sha256::Hash>,
        acc: Stump,
    ) -> Result<(), Self::Error> {
        // verify the proof
        let del_hashes = del_hashes
            .iter()
            .map(|hash| NodeHash::from(hash.as_byte_array()))
            .collect::<Vec<_>>();

        if !acc.verify(&proof, &del_hashes)? {
            return Err(BlockValidationErrors::InvalidProof.into());
        }

        let height = self
            .get_block_height(&block.block_hash())?
            .ok_or(BlockchainError::BlockNotPresent)?;

        self.validate_block(block, height, inputs)
    }
```
but we are executing the `validate_block` function in the `ChainState` trait in `connect_block` function. 

@Davidson-Souza

Reply:
> Should I let it be and change the comments accordingly or ... ??

Yes, please

> but we are executing the validate_block function in the ChainState trait in connect_block function. So where will the InvalidProof error message come from if not from BlokchainInterface ??

Invalid proof comes from `connect_block`, but we can't invalidate the block because of an invalid proof. This is due to proofs not being committed in the block itself, so a dishonest peer can produce an invalid proof without invalidating the PoW and make us invalidate that block, throwing us out of the network. Maybe you can add a test with this case: Some peer sends a block with invalid proof, we ban this peer and ask another. The second peer gives us a valid proof and we accept that block.